### ISSUE TITLE
[Graphbolt] Performances of various UVA index select kernels.

### ISSUE THREAD

**Issue by czkkkkkk** [dmlc/dgl#6590]
## Description

There are several factors to decide in the current UVA index select kernel, including alignment, index permutation, index sorting, and the specialization for single feature. In this issue, I profiled the contributions of these factors to the optimal performance. From the results of both random and real datasets on a g4dn.4xlarge instance (a single T4 GPU), it seems that we only needs to keep an aligned access kernel without permutation and specialization to ensure nearly the optimal performance. The profiling code can be found in #6589.

@mfbalin, could you take a look at the setting to see whether the comparisons are fair? Thanks!

## Setting
1. Datasets
    * Random: randomly generate a `n_rows` table with `feat_size` columns and a list of `num_indices` indices to access the table.
    * Real: apply neighbor sampling on `ogbn-arxiv` or `ogbn-products` and generate a list of `indices` to access the corresponding features.
2. Kernel options
    * USE_PERM: whether sort the indices before index select
    * USE_SINGLE: whether use the specialized kernel when `feat_size=1`
    * USE_TORCH_SORT: If `USE_PERM=1`, it decides whether use pytorch or cub for index sorting.
    * USE_ALIGN: whether use aligned kernel access kernel.

## Results

The performance metric is the throughput of fetching data through UVA (MB/s). As a reference, the peak PCIe bandwidth for copying data from CPU to GPU is 6.6 GB/s. If some options are not tested, their default values are `USE_PERM=1`, `USE_SINGLE=1`, `USE_TORCH_SORT=0`, and `USE_ALIGN=1`.

### Conclusions

* Specialized kernel for `feat_size=1` has a similar performance to unspecialized kernels.
* Sorting the indices could slow down the performance. Even cub sorting is faster than torch sort, sorting is also more expensive than the saving from aligned memory access.
* Aligned memory access is slightly faster than unaligned kernels.
### Testing `USE_PERM` and `USE_SINGLE` on the random datasets.

* params:  n_rows=100000, feat_size=1, num_indices=1, feature_device=Device.Pinned, indices_device=Device.GPU

|            |USE_SINGLE=1|USE_SINGLE=0|
|------------|------------|------------|
|USE_PERM=1  |       0.044|       0.043|
|USE_PERM=0  |       0.121|       0.114|

* params:  n_rows=100000, feat_size=1, num_indices=1000, feature_device=Device.Pinned, indices_device=Device.GPU

|            |USE_SINGLE=1|USE_SINGLE=0|
|------------|------------|------------|
|USE_PERM=1  |      37.587|      31.051|
|USE_PERM=0  |      92.894|      64.940|

* params:  n_rows=100000, feat_size=1, num_indices=10000, feature_device=Device.Pinned, indices_device=Device.GPU

|            |USE_SINGLE=1|USE_SINGLE=0|
|------------|------------|------------|
|USE_PERM=1  |     140.430|     126.677|
|USE_PERM=0  |     290.893|     332.026|

### Testing `USE_PERM` and `USE_TORCH_SORT` on random datasets.

* params:  n_rows=100000, feat_size=100, num_indices=1, feature_device=Device.Pinned, indices_device=Device.GPU

|            |USE_TORCH_SORT=1|USE_TORCH_SORT=0|
|------------|------------|------------|
|USE_PERM=1  |       3.364|       4.454|
|USE_PERM=0  |      11.869|      11.901|
* params:  n_rows=100000, feat_size=100, num_indices=1000, feature_device=Device.Pinned, indices_device=Device.GPU

|            |USE_TORCH_SORT=1|USE_TORCH_SORT=0|
|------------|------------|------------|
|USE_PERM=1  |    1605.974|    2421.417|
|USE_PERM=0  |    3715.314|    3718.855|
* params:  n_rows=100000, feat_size=100, num_indices=10000, feature_device=Device.Pinned, indices_device=Device.GPU

|            |USE_TORCH_SORT=1|USE_TORCH_SORT=0|
|------------|------------|------------|
|USE_PERM=1  |    3983.310|    4744.072|
|USE_PERM=0  |    5172.771|    5233.496|
* params:  n_rows=100000, feat_size=300, num_indices=1, feature_device=Device.Pinned, indices_device=Device.GPU

|            |USE_TORCH_SORT=1|USE_TORCH_SORT=0|
|------------|------------|------------|
|USE_PERM=1  |      11.516|      15.274|
|USE_PERM=0  |      38.023|      38.278|
* params:  n_rows=100000, feat_size=300, num_indices=1000, feature_device=Device.Pinned, indices_device=Device.GPU

|            |USE_TORCH_SORT=1|USE_TORCH_SORT=0|
|------------|------------|------------|
|USE_PERM=1  |    3498.823|    4165.763|
|USE_PERM=0  |    4548.814|    4778.735|
* params:  n_rows=100000, feat_size=300, num_indices=10000, feature_device=Device.Pinned, indices_device=Device.GPU

|            |USE_TORCH_SORT=1|USE_TORCH_SORT=0|
|------------|------------|------------|
|USE_PERM=1  |    5421.254|    5651.394|
|USE_PERM=0  |    5686.685|    5677.013|

#### Testing `USE_PERM` and `USE_ALIGN`

* params:  n_rows=100000, feat_size=4, num_indices=1, feature_device=Device.Pinned, indices_device=Device.GPU

|            |USE_ALIGN=1 |USE_ALIGN=0 |
|------------|------------|------------|
|USE_PERM=1  |       0.182|       0.180|
|USE_PERM=0  |       0.451|       0.460|
* params:  n_rows=100000, feat_size=4, num_indices=1000, feature_device=Device.Pinned, indices_device=Device.GPU

|            |USE_ALIGN=1 |USE_ALIGN=0 |
|------------|------------|------------|
|USE_PERM=1  |     150.129|     150.936|
|USE_PERM=0  |     356.228|     369.771|
* params:  n_rows=100000, feat_size=4, num_indices=10000, feature_device=Device.Pinned, indices_device=Device.GPU

|            |USE_ALIGN=1 |USE_ALIGN=0 |
|------------|------------|------------|
|USE_PERM=1  |     601.063|     616.321|
|USE_PERM=0  |    1313.899|    1319.860|
* params:  n_rows=100000, feat_size=16, num_indices=1, feature_device=Device.Pinned, indices_device=Device.GPU

|            |USE_ALIGN=1 |USE_ALIGN=0 |
|------------|------------|------------|
|USE_PERM=1  |       0.674|       0.677|
|USE_PERM=0  |       1.868|       1.871|
* params:  n_rows=100000, feat_size=16, num_indices=1000, feature_device=Device.Pinned, indices_device=Device.GPU

|            |USE_ALIGN=1 |USE_ALIGN=0 |
|------------|------------|------------|
|USE_PERM=1  |     585.678|     582.825|
|USE_PERM=0  |    1332.223|    1344.397|
* params:  n_rows=100000, feat_size=16, num_indices=10000, feature_device=Device.Pinned, indices_device=Device.GPU

|            |USE_ALIGN=1 |USE_ALIGN=0 |
|------------|------------|------------|
|USE_PERM=1  |    2061.590|    2075.698|
|USE_PERM=0  |    3881.022|    3880.199|

* params:  n_rows=100000, feat_size=100, num_indices=1, feature_device=Device.Pinned, indices_device=Device.GPU

|            |USE_ALIGN=1 |USE_ALIGN=0 |
|------------|------------|------------|
|USE_PERM=1  |       4.535|       4.540|
|USE_PERM=0  |      12.051|      12.075|
* params:  n_rows=100000, feat_size=100, num_indices=1000, feature_device=Device.Pinned, indices_device=Device.GPU

|            |USE_ALIGN=1 |USE_ALIGN=0 |
|------------|------------|------------|
|USE_PERM=1  |    2427.847|    2426.007|
|USE_PERM=0  |    3706.535|    3715.573|
* params:  n_rows=100000, feat_size=100, num_indices=10000, feature_device=Device.Pinned, indices_device=Device.GPU

|            |USE_ALIGN=1 |USE_ALIGN=0 |
|------------|------------|------------|
|USE_PERM=1  |    4666.886|    4767.140|
|USE_PERM=0  |    5228.143|    5229.220|
* params:  n_rows=100000, feat_size=300, num_indices=1, feature_device=Device.Pinned, indices_device=Device.GPU

|            |USE_ALIGN=1 |USE_ALIGN=0 |
|------------|------------|------------|
|USE_PERM=1  |      15.052|      15.215|
|USE_PERM=0  |      38.253|      38.375|
* params:  n_rows=100000, feat_size=300, num_indices=1000, feature_device=Device.Pinned, indices_device=Device.GPU

|            |USE_ALIGN=1 |USE_ALIGN=0 |
|------------|------------|------------|
|USE_PERM=1  |    4142.359|    4043.808|
|USE_PERM=0  |    4604.979|    4408.078|
* params:  n_rows=100000, feat_size=300, num_indices=10000, feature_device=Device.Pinned, indices_device=Device.GPU

|            |USE_ALIGN=1 |USE_ALIGN=0 |
|------------|------------|------------|
|USE_PERM=1  |    5667.502|    5664.399|
|USE_PERM=0  |    5685.601|    5675.517|

### Testing on ogbn-arxiv and products.

* params:  dataset=ogbn-arxiv n_rows=169343 feat_size=128 num_indices=89 feature_device=Device.Pinned, indices_device=Device.GPU

|            |USE_TORCH_SORT=1|USE_TORCH_SORT=0|
|------------|------------|------------|
|USE_PERM=1  |    5325.152|    5574.543|
|USE_PERM=0  |    5874.786|    5878.957|

* params:  dataset=ogbn-arxiv n_rows=169343 feat_size=128 num_indices=89 feature_device=Device.Pinned, indices_device=Device.GPU

|            |USE_ALIGN=1 |USE_ALIGN=0 |
|------------|------------|------------|
|USE_PERM=1  |    5571.091|    5580.243|
|USE_PERM=0  |    5877.256|    5876.618|

* params:  dataset=ogbn-products n_rows=2449029 feat_size=100 num_indices=193 feature_device=Device.Pinned, indices_device=Device.GPU

|            |USE_TORCH_SORT=1|USE_TORCH_SORT=0|
|------------|------------|------------|
|USE_PERM=1  |    5451.208|    5529.966|
|USE_PERM=0  |    5466.762|    5465.928|

* params:  dataset=ogbn-products n_rows=2449029 feat_size=100 num_indices=193 feature_device=Device.Pinned, indices_device=Device.GPU

|            |USE_ALIGN=1 |USE_ALIGN=0 |
|------------|------------|------------|
|USE_PERM=1  |    5528.438|    5528.021|
|USE_PERM=0  |    5465.073|    5464.857|

**Comment by mfbalin** [comment]
@czkkkkkk thank you for the investigation. The sorting optimization PR was motivated by the lack of performance on the mag240M dataset. Can you test with the mag240M scenarios below:
1. n_rows = 3B, feat_size=1, num_indices=1M for edge type accesses with int8 datatype.
2. n_rows = 240M, feat_size=768, num_indices=1M for vertex feature accesses with float16 dtype.

Performance matters a lot more for the large scale training scenario.

**Comment by mfbalin** [comment]
Resource usage can be measured by looking at the profile and recording how many threadblocks were used.

**Comment by mfbalin** [comment]
@czkkkkkk My expected conclusions with the above considerations would be as follows:

* Specialized kernel for feat_size=1 has a similar performance to unspecialized kernels. However, it can reduce the number of threadblocks used significantly.
* Sorting the indices could slow down the performance. But for large-scale scenarios, it will make a big performance difference.
* Aligned memory access is slightly faster than unaligned kernels. And this performance difference is expected to grow for the larger scale use cases, especially if feat_size is not a good number. To test this, we can try a prime number feat_size.

**Comment by mfbalin** [comment]
I believe the regression tests and optimized version of the kernel has been implemented.

Feel free to open the issue if you think there is a way to improve it so that it won't cause a regression.
